File: src/dart/lib/src/script/dart_resource_format.dart
import 'dart:ffi';

import 'package:path/path.dart' as p;

import '../../godot_dart.dart';
import 'dart_script.dart';

class DartResourceFormatLoader extends ResourceFormatLoader {
  static TypeInfo sTypeInfo = TypeInfo(
    DartResourceFormatLoader,
    StringName.fromString('DartResourceFormatLoader'),
    parentClass: ResourceFormatLoader.sTypeInfo.className,
  );
  static Map<String, Pointer<GodotVirtualFunction>> get vTable =>
      ResourceFormatLoader.vTable;

  DartResourceFormatLoader() : super() {
    postInitialize();
  }

  static void initBindings() {
    gde.dartBindings.bindClass(
        DartResourceFormatLoader, DartResourceFormatLoader.sTypeInfo);
  }

  @override
  TypeInfo get typeInfo => sTypeInfo;

  @override
  bool vHandlesType(String type) {
    return type == 'Script' || type == 'DartScript';
  }

  @override
  PackedStringArray vGetRecognizedExtensions() {
    final array = PackedStringArray();
    array.pushBack('dart');
    return array;
  }

  @override
  bool vRecognizePath(String path, String type) {
    return path.endsWith('.dart');
  }

  @override
  String vGetResourceType(String path) {
    final extension = p.extension(path);

    return extension == '.dart' ? 'DartScript' : '';
  }

  @override
  String vGetResourceScriptClass(String path) {
    return 'DartScript';
  }

  @override
  bool vExists(String path) {
    return FileAccess.fileExists(path);
  }

  @override
  Variant vLoad(
      String path, String originalPath, bool useSubThreads, int cacheMode) {
    // Can cast directly since this is a direct Dart -> Dart call
    final script = DartScriptLanguage.singleton.vCreateScript() as DartScript?;

    if (script == null) {
      return Variant();
    }

    script.setPath(originalPath);
    final file = FileAccess.open(originalPath, FileAccessModeFlags.read);
    if (file.obj != null) {
      final text = file.obj?.getAsText(false);
      if (text != null) {
        script.setSourceCode(text);
      }
      file.obj?.close();
    }

    return convertToVariant(script);
  }

  @override
  int vGetResourceUid(String path) {
    return -1;
  }
}

class DartResourceFormatSaver extends ResourceFormatSaver {
  static TypeInfo sTypeInfo = TypeInfo(
    DartResourceFormatSaver,
    StringName.fromString('DartResourceFormatSaver'),
    parentClass: ResourceFormatSaver.sTypeInfo.className,
  );
  static Map<String, Pointer<GodotVirtualFunction>> get vTable =>
      ResourceFormatSaver.vTable;

  DartResourceFormatSaver() : super() {
    postInitialize();
  }

  static void initBindings() {
    gde.dartBindings
        .bindClass(DartResourceFormatSaver, DartResourceFormatSaver.sTypeInfo);
  }

  @override
  TypeInfo get typeInfo => sTypeInfo;

  @override
  GDError vSave(Ref<Resource> resource, String path, int flags) {
    if (resource.obj == null) return GDError.errInvalidParameter;

    final script = gde.cast<DartScript>(resource.obj, DartScript.sTypeInfo);
    if (script == null) return GDError.errBug;

    final file = FileAccess.open(path, FileAccessModeFlags.write);
    if (file.obj != null) {
      final srcCode = script.getSourceCode();
      file.obj?.storeString(srcCode);
      if (file.obj?.getError() != GDError.ok &&
          file.obj?.getError() != GDError.errFileEof) {
        return GDError.errCantCreate;
      }
    }

    file.obj?.flush();
    file.obj?.close();

    return GDError.ok;
  }

  @override
  bool vRecognizePath(Ref<Resource> resource, String path) {
    return path.endsWith('.dart');
  }

  @override
  bool vRecognize(Ref<Resource> resource) {
    if (gde.cast<DartScript>(resource.obj, DartScript.sTypeInfo) != null) {
      return true;
    }
    return false;
  }

  @override
  PackedStringArray vGetRecognizedExtensions(Ref<Resource> resource) {
    return PackedStringArray()..pushBack('dart');
  }
}

File: src/cpp/script/dart_resource_format.cpp
#include "dart_resource_format.h"

#include <godot_cpp/classes/resource_loader.hpp>
#include <godot_cpp/classes/file_access.hpp>

#include "dart_script_language.h"
#include "dart_script.h"

using namespace godot;

// ResourceFormatLoader

DartResourceFormatLoader::DartResourceFormatLoader() {

}

void DartResourceFormatLoader::_bind_methods() {
}

bool DartResourceFormatLoader::_handles_type(const godot::StringName &type) const {
  return type == StringName("Script") || type == StringName("DartScript");
}

godot::PackedStringArray DartResourceFormatLoader::_get_recognized_extensions() const {
  PackedStringArray array;
  array.push_back("dart");
  return array;
}

bool DartResourceFormatLoader::_recognize_path(const godot::String &path, const godot::StringName &type) const {
  return path.ends_with(".dart");
}

godot::String DartResourceFormatLoader::_get_resource_type(const godot::String &path) const {
  String extension = path.get_extension();

  return extension == String(".dart") ? String("DartScript") : String();
}

godot::String DartResourceFormatLoader::_get_resource_script_class(const godot::String &path) const {
  return String("DartScript");
}

bool DartResourceFormatLoader::_exists(const godot::String &path) const {
  return FileAccess::file_exists(path);
}

godot::Variant DartResourceFormatLoader::_load(const godot::String &path, const godot::String &original_path,
                                               bool use_sub_threads, int32_t cache_mode) const {
  DartScriptLanguage *language = DartScriptLanguage::instance();
  Ref<DartScript> script = language->get_cached_script(path);
  if (script.is_null()) {
    script = Ref<DartScript>(language->_create_script());
    if (script.is_null()) {
      return Variant();
    }

    language->push_cached_script(path, script);
    script->load_from_disk(original_path);
  } else if (cache_mode == ResourceLoader::CACHE_MODE_IGNORE) {
    script->load_from_disk(original_path);
  }
  script->set_path(original_path);

  return Variant(script.ptr());
}

// ResourceFormatSaver

void DartResourceFormatSaver::_bind_methods() {
}


godot::Error DartResourceFormatSaver::_save(const godot::Ref<godot::Resource> &resource, const godot::String &path,
                                            uint32_t flags) {
  Ref<DartScript> script = Object::cast_to<DartScript>(resource.ptr());
  if (script.is_null()) {
    return ERR_BUG;
  }

  Ref<FileAccess> file = FileAccess::open(path, FileAccess::ModeFlags::WRITE);
  if (file.is_null()) {
    return ERR_BUG;
  }

  String source = script->get_source_code();
  file->store_string(source);
  if (file->get_error() != OK && file->get_error() != ERR_FILE_EOF) {
    return ERR_CANT_CREATE;
  }

  file->flush();
  file->close();

  return OK;
}

bool DartResourceFormatSaver::_recognize(const godot::Ref<godot::Resource> &resource) const {
  DartScript* script = Object::cast_to<DartScript>(resource.ptr());
  return script != nullptr;
}

bool DartResourceFormatSaver::_recognize_path(const godot::Ref<godot::Resource> &resource,
                                              const godot::String &path) const {
  return path.ends_with(".dart");
}

godot::PackedStringArray DartResourceFormatSaver::_get_recognized_extensions(
    const godot::Ref<godot::Resource> &resource) const {
  PackedStringArray array;
  array.push_back("dart");
  return array;
}
